fees/royco.ts

```diff
@@ -0,0 +1,236 @@
+        const event: RecipeEvent = {
+          offerHash: item.offerHash,
+          fillAmount: item.fillAmount,
+          incentiveAmounts: item.incentiveAmounts,
+          protocolFeeAmounts: item.protocolFeeAmounts,
+          frontendFeeAmounts: item.frontendFeeAmounts,
+
+          // will fill later
+          incentiveTokens: [],
+        }
+
+        receiptEvents.push(event)
+        startOfferHash = event.offerHash
+      }
+
+      if (response.ipofferFilleds.length === 0) {
+        // break loop
+        startOfferHash = ''
+      }
+    } while(startOfferHash !== '')
+
+    if (receiptEvents.length === 0) {
+      return;
+    }
+
+    // query marketId of given offerHash
+    const offerHashList: Array<string> = receiptEvents.map(item => item.offerHash)
+    const query_ipofferCreateds = gql`
+      query get_ipofferCreateds($offers: [String]) {
+        ipofferCreateds(first: ${offerHashList.length}, where: {offerHash_in: $offers}) {
+          offerHash
+          marketHash
+        }
+      }
+    `
+    const query_ipofferCreatedsResponse = await request(endpoint, query_ipofferCreateds, {
+      offers: offerHashList,
+    })
+
+    // query incentive tokens of given marketId
+    const query_rawMarkets = gql`
+      query get_rawMarkets($markets: [String]) {
+        rawMarkets(first: ${offerHashList.length}, where: {marketId_in: $markets}) {
+          marketId
+          incentivesOfferedIds
+        }
+      }
+    `
+    const query_rawMarketsResponse = await request(endpoint, query_rawMarkets, {
+      markets: query_ipofferCreatedsResponse.ipofferCreateds.map((item: any) => item.marketHash),
+    })
+
+    // map offer offerHash -> raw market id
+    const offerHashToIncentiveTokens: {[key: string]: Array<string>} = {}
+    for (const offer of query_ipofferCreatedsResponse.ipofferCreateds) {
+      const rawMarket = query_rawMarketsResponse.rawMarkets.find((item: any) => item.marketId === offer.marketHash)
+      if (rawMarket) {
+        // parse incentive tokens from format of: chain-address
+        // 1-0x6243558a24cc6116abe751f27e6d7ede50abfc76
+        offerHashToIncentiveTokens[offer.offerHash] = rawMarket.incentivesOfferedIds.map((token: string) => token.split('-')[1])
+      }
+    }
+
+    // update event incentive tokens
+    for (let i = 0; i < receiptEvents.length; i++) {
+      receiptEvents[i].incentiveTokens = offerHashToIncentiveTokens[receiptEvents[i].offerHash]
+    }
+
+    for (const event of receiptEvents) {
+      for (let i = 0; i < event.incentiveTokens.length; i++) {
+        dailySupplySideRevenue.add(event.incentiveTokens[i], event.incentiveAmounts[i])
+        dailyProtocolRevenue.add(event.incentiveTokens[i], event.protocolFeeAmounts[i])
+        dailyProtocolRevenue.add(event.incentiveTokens[i], event.frontendFeeAmounts[i])
+      }
+    }
+}
+
+function getFetchFunction(defillamaChain: string): FetchV2 {
+  return async function (options: FetchOptions): Promise<FetchResultV2> {
```

Comment:
you can pull chain string from `options.chain` , so this wrapped is not needed, with the wrapper, all adapter objects would be the same, then you can do something like:

    const _adapter ={
      fetch,
      meta: {
        methodology,
      },
      start: '2024-12-02',
    }

    const adapter: Adapter = {
      version: 2,
      adapter: {
        [CHAIN.ETHEREUM]: _adapter,
        [CHAIN.ARBITRUM]: _adapter,
        [CHAIN.BASE]: _adapter,
        [CHAIN.CORN]: _adapter,
        [CHAIN.SONIC]: _adapter,
        [CHAIN.HYPERLIQUID]: _adapter,
      }
    }